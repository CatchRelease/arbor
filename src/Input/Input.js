import React from 'react';
import PropTypes from 'prop-types';

import FormField from '../FormField';
import StyledInput from './StyledInput';

const Input = React.forwardRef(
  ({ caption, label, labelAside, id, ...props }, ref) => (
    <FormField caption={caption} id={id} label={label} labelAside={labelAside}>
      <StyledInput {...{ ...props, id, ref }} />
    </FormField>
  )
);

Input.propTypes = {
  /**
   * The actualy HTML element tag to use when rendering.
   */
  as: PropTypes.string,

  /**
   * Optional text to be displayed underneath the input.
   */
  caption: PropTypes.string,

  /**
   * HTML ID for the input to be used in conjunction with the label for
   * accessiblity.
   */
  id: PropTypes.string.isRequired,

  /**
   * Label text for the form field
   */
  label: PropTypes.string,

  /**
   * Component that goes to the right of the label. Does not have to only be text.
   */
  labelAside: PropTypes.node,

  /**
   * HTML Input type
   */
  type: PropTypes.string
};

Input.defaultProps = {
  as: 'input',
  caption: '',
  label: '',
  labelAside: null,
  type: 'text'
};

export default Input;
