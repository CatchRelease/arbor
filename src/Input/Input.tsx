import {
  ChangeEventHandler,
  FC,
  FocusEventHandler,
  forwardRef,
  InputHTMLAttributes,
  ReactNode,
  useState
} from 'react';

import FormField from '../FormField';
import StyledInput, { Props as StyledInputProps } from './StyledInput';

type Props = InputHTMLAttributes<HTMLInputElement> &
  StyledInputProps & {
    caption?: string | null;
    id: string;
    label?: string;
    labelAside?: ReactNode;
    onBlur?: FocusEventHandler<HTMLInputElement>;
    onChange?: ChangeEventHandler<HTMLInputElement>;
    type?: string;
    validate?: (value?: string | number | readonly string[]) => string | null;
  };

const Input: FC<Props> = forwardRef<HTMLInputElement, Props>(
  (
    {
      caption = '',
      label = '',
      labelAside = null,
      id,
      onBlur,
      onChange,
      type = 'text',
      validate = () => undefined,
      value: propValue,
      ...props
    },
    ref
  ) => {
    const [touched, setTouched] = useState(false);
    const [value, setValue] = useState(propValue);

    const errorMessage = touched && validate(value);

    const handleBlur: FocusEventHandler<HTMLInputElement> = (event) => {
      setTouched(true);

      if (onBlur) {
        onBlur(event);
      }
    };

    const handleChange: ChangeEventHandler<HTMLInputElement> = (event) => {
      setValue(event.target.value);

      if (onChange) {
        onChange(event);
      }
    };

    return (
      <FormField
        caption={errorMessage || caption}
        id={id}
        label={label}
        labelAside={labelAside}
      >
        <StyledInput
          {...{ ...props, id, ref, type }}
          value={propValue ? propValue.toString() : propValue}
          isInvalid={!!errorMessage}
          onBlur={handleBlur}
          onChange={handleChange}
        />
      </FormField>
    );
  }
);

export default Input;
