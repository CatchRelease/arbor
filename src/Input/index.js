import React from 'react';
import PropTypes from 'prop-types';

import InputContainer from './InputContainer';
import StyledInput from './StyledInput';
import Label from '../Label';
import Paragraph from '../Paragraph';

const Input = ({ caption, label, id, ...props }) => (
  <InputContainer caption={caption}>
    <Label htmlFor={id}>{label}</Label>
    <StyledInput {...{ ...props, id }} />
    {caption && (
      <Paragraph.span variant="tiny" color="red">
        {caption}
      </Paragraph.span>
    )}
  </InputContainer>
);

Input.propTypes = {
  caption: PropTypes.string,
  id: PropTypes.string.isRequired,
  large: PropTypes.bool
};

Input.defaultProps = {
  caption: '',
  large: false
};

export default Input;
