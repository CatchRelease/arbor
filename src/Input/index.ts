export { default, Props as InputProps } from './Input';
export { default as inputStyles } from './inputStyles';
