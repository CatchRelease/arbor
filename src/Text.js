import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { color, fontSize, space, textAlign } from 'styled-system';

const baseStyles = ({ theme }) => css`
  font-family: ${theme.brandFont};
  line-height: ${theme.lineHeights.large};
`;

const Text = styled.p`
  ${color};
  ${fontSize};
  ${space};
  ${textAlign};
  ${baseStyles};
`;

Text.propTypes = {
  ...color.propTypes,
  ...fontSize.propTypes,
  ...space.propTypes,
  ...textAlign.propTypes
};

Text.p = Text.withComponent('p');
Text.span = Text.withComponent('span');

export default Text;
