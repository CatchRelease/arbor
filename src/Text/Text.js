import React from 'react';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import {
  color,
  fontSize,
  fontWeight,
  lineHeight,
  maxWidth,
  minWidth,
  overflow,
  space,
  textAlign
} from 'styled-system';

import textTransform from '../utils/textTransform';
import textOverflow from '../utils/textOverflow';
import whiteSpace from '../utils/whiteSpace';

const baseStyles = ({ theme }) => css`
  font-family: ${theme.brandFont};
`;

const Text = styled.p`
  ${baseStyles};

  ${color};
  ${fontSize};
  ${fontWeight};
  ${lineHeight};
  ${maxWidth};
  ${minWidth};
  ${overflow};
  ${space};
  ${textAlign};
  ${textTransform};
  ${textOverflow};
  ${whiteSpace};
`;

Text.propTypes = {
  ...color.propTypes,
  ...fontSize.propTypes,
  ...fontWeight.propTypes,
  ...lineHeight.propTypes,
  ...maxWidth.propTypes,
  ...minWidth.propTypes,
  ...overflow.propTypes,
  ...space.propTypes,
  ...textAlign.propTypes,
  ...textTransform.propTypes,
  ...textOverflow.propTypes,
  ...whiteSpace.propTypes
};

Text.defaultProps = {
  color: 'text.default',
  fontSize: 'size4',
  lineHeight: 'small'
};

// Deprecated -- use <Text as="span"></Text>
Text.Span = (props) => <Text as="span" {...props} />;

export default Text;
