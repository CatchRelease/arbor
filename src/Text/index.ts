export { default, Props as TextProps } from './Text';
