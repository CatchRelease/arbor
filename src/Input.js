import React from 'react';
import styled, { css } from 'react-emotion';
import PropTypes from 'prop-types';
import { placeholder, transparentize } from 'polished';

import {
  borderRadius,
  borderWidth,
  colors,
  fontSizes,
  lineHeights,
  spacings
} from './theme';

import Label from './Label';
import Text from './Text';

const InputContainer = styled.div`
  margin-bottom: ${props =>
    props.caption ? spacings.smaller : spacings.regular};
`;

InputContainer.propTypes = {
  caption: PropTypes.string
};

InputContainer.defaultProps = {
  caption: ''
};

const largeStyles = ({ large }) =>
  large &&
  css`
    padding: ${spacings.small} ${spacings.smaller};
  `;

const StyledInput = styled.input`
  border-radius: ${borderRadius.small};
  border: ${borderWidth.small} solid ${colors.grey20};
  box-sizing: border-box;
  color: ${colors.grey100};
  font-size: ${fontSizes.size2};
  line-height: ${lineHeights.small};
  margin: ${spacings.smallest} 0;
  padding: ${spacings.smaller};
  width: 100%;
  ${largeStyles};

  ${placeholder({ color: colors.grey50 })};

  &:focus {
    border: ${borderWidth.small} solid ${colors.blueLight};
    box-shadow: 0 0 8px ${transparentize(0.9, colors.blueLight)};
  }

  &:disabled {
    background: ${colors.white10};
  }
`;

const Input = ({ caption, label, id, ...props }) => (
  <InputContainer caption={caption}>
    <Label htmlFor={id}>{label}</Label>
    <StyledInput {...{ ...props, id }} />
    {caption && (
      <Text.span variant="tiny" color="red" withoutMargin>
        {caption}
      </Text.span>
    )}
  </InputContainer>
);

Input.propTypes = {
  caption: PropTypes.string,
  id: PropTypes.string.isRequired,
  large: PropTypes.bool
};

Input.defaultProps = {
  caption: '',
  large: false
};

export default Input;
