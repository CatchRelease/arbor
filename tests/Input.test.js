import React from 'react';
import { shallow } from 'enzyme';
import { create } from 'react-test-renderer';

import Input from '../src/Input';
import Text from '../src/Text';

describe('<Input />', () => {
  it('renders an Input correctly', () => {
    const tree = create(
      <Input label="My Input" id="example" type="text" />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('associates label with input', () => {
    const input = shallow(<Input type="text" id="example" />);

    expect(input.find('Label').props().htmlFor).toEqual('example');
  });

  describe('caption exists', () => {
    it('renders the caption', () => {
      const input = shallow(<Input caption="My Caption" id="example" />);

      expect(input.find(Text.span).props().children).toEqual('My Caption');
    });
  });

  describe('caption does not exist', () => {
    it('does not render the caption', () => {
      const input = shallow(<Input id="example" />);

      expect(input.find(Text.span).length).toBe(0);
    });
  });
});
