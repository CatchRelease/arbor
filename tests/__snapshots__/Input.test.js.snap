// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> renders a large Input correctly 1`] = `
.emotion-4 {
  margin-bottom: 1rem;
}

.emotion-0 {
  color: #666666;
  display: block;
  font-family: Open Sans;
  font-size: 0.875rem;
  line-height: 1.5;
}

.emotion-2 {
  border-radius: 3px;
  border: 1px solid #DDDDDD;
  box-sizing: border-box;
  color: #222222;
  font-size: 0.875rem;
  line-height: 1.25;
  margin: 0.25rem 0;
  padding: 0.5rem;
  width: 100%;
  padding: 0.75rem 0.5rem;
}

.emotion-2::-webkit-input-placeholder {
  color: #AAAAAA;
}

.emotion-2:-moz-placeholder {
  color: #AAAAAA;
}

.emotion-2::-moz-placeholder {
  color: #AAAAAA;
}

.emotion-2:-ms-input-placeholder {
  color: #AAAAAA;
}

.emotion-2:focus {
  border: 1px solid #5577FF;
  box-shadow: 0 0 8px rgba(85,119,255,0.1);
}

.emotion-2:disabled {
  background: #FAFAFA;
}

<div
  className="emotion-4 emotion-5"
>
  <label
    className="emotion-0 emotion-1"
    htmlFor="example"
  >
    My Input
  </label>
  <input
    className="emotion-2 emotion-3"
    id="example"
    type="text"
  />
  
</div>
`;

exports[`<Input /> renders an Input correctly 1`] = `
.emotion-4 {
  margin-bottom: 1rem;
}

.emotion-0 {
  color: #666666;
  display: block;
  font-family: Open Sans;
  font-size: 0.875rem;
  line-height: 1.5;
}

.emotion-2 {
  border-radius: 3px;
  border: 1px solid #DDDDDD;
  box-sizing: border-box;
  color: #222222;
  font-size: 0.875rem;
  line-height: 1.25;
  margin: 0.25rem 0;
  padding: 0.5rem;
  width: 100%;
}

.emotion-2::-webkit-input-placeholder {
  color: #AAAAAA;
}

.emotion-2:-moz-placeholder {
  color: #AAAAAA;
}

.emotion-2::-moz-placeholder {
  color: #AAAAAA;
}

.emotion-2:-ms-input-placeholder {
  color: #AAAAAA;
}

.emotion-2:focus {
  border: 1px solid #5577FF;
  box-shadow: 0 0 8px rgba(85,119,255,0.1);
}

.emotion-2:disabled {
  background: #FAFAFA;
}

<div
  className="emotion-4 emotion-5"
>
  <label
    className="emotion-0 emotion-1"
    htmlFor="example"
  >
    My Input
  </label>
  <input
    className="emotion-2 emotion-3"
    id="example"
    type="text"
  />
  
</div>
`;

exports[`<Input /> renders an Input with a caption correctly 1`] = `
.emotion-0 {
  color: #666666;
  display: block;
  font-family: Open Sans;
  font-size: 0.875rem;
  line-height: 1.5;
}

.emotion-2 {
  border-radius: 3px;
  border: 1px solid #DDDDDD;
  box-sizing: border-box;
  color: #222222;
  font-size: 0.875rem;
  line-height: 1.25;
  margin: 0.25rem 0;
  padding: 0.5rem;
  width: 100%;
}

.emotion-2::-webkit-input-placeholder {
  color: #AAAAAA;
}

.emotion-2:-moz-placeholder {
  color: #AAAAAA;
}

.emotion-2::-moz-placeholder {
  color: #AAAAAA;
}

.emotion-2:-ms-input-placeholder {
  color: #AAAAAA;
}

.emotion-2:focus {
  border: 1px solid #5577FF;
  box-shadow: 0 0 8px rgba(85,119,255,0.1);
  outline: none;
}

.emotion-2:disabled {
  background: #FAFAFA;
}

.emotion-6 {
  margin-bottom: 0.5rem;
}

.emotion-4 {
  font-family: Open Sans;
  line-height: 1.5;
  color: #EE2200;
  font-size: 0.6875rem;
  margin-bottom: 0.6875rem;
  margin-bottom: 0;
}

<div
  className="emotion-6 emotion-7"
>
  <label
    className="emotion-0 emotion-1"
    htmlFor="example"
  >
    My Input
  </label>
  <input
    className="emotion-2 emotion-3"
    id="example"
    type="text"
  />
  <span
    className="emotion-4 emotion-5"
    color="red"
  >
    I am a caption
  </span>
</div>
`;
